### ISSUE TITLE
Support proxy in param

### ISSUE THREAD

**Issue by tianfeng92** [microsoft/playwright-test#257]
Hi there,
As a user, I'd like to have a way to set proxy via `folio` with `playwright-test`. It'd be better to add a field in `param` named `proxy` so that I can set proxy by running
```
npx folio --param proxy="http://my-proxy"
```
Some tests need to access some internal websites and it requires some proxy settings. Without that most of the cases cannot be covered. Currently, I didn't find a way to pass `proxy` to `playwright-test`. I've seen `playwright` supporting proxy in launch option so it'd nice to get that supported in `folio`/`playwright-test` as well.

Thanks!
Tian

**Comment by mxschmitt** [comment]
Normally if the proxy is configured locally via the `http_proxy` and `https_proxy` environment variable, then the browsers should pick them up automatically.

**Comment by mxschmitt** [comment]
Could you try setting it in lower-case like `http_proxy`. Firefox as far as I know unfortunately only looks for the lower-case variant.

**Comment by alexplischke** [comment]
Hope this provides a bit more insight (seems to be HTTPS related) :)

1. Setting proxy env variables (some bogus proxy, so I expect the test to fail)
```
~> export HTTP_PROXY=http://localhost:1234
~> export HTTPS_PROXY=https://localhost:1234
```

2. Running playwright via folio (against HTTP)
```
~> npx folio --param browserName=chromium

Running 1 test using 1 worker
  1) tests/example.test.js:3:1 › should verify title of the page ===================================
     browserName=chromium, headful=false, slowMo=0, video=false, screenshotOnFailure=false

    page.goto: net::ERR_NAME_NOT_RESOLVED at http://whatsmyuseragent.org/
    =========================== logs ===========================
    navigating to "http://whatsmyuseragent.org/", waiting until "load"
    ============================================================
    [...]
```
The test fails as expected.

3. Running playwright directly (against HTTP)
```
~> node direct.js

(node:24351) UnhandledPromiseRejectionWarning: page.goto: net::ERR_NAME_NOT_RESOLVED at http://whatsmyuseragent.org/
=========================== logs ===========================
navigating to "http://whatsmyuseragent.org/", waiting until "load"
============================================================
[...]
```
Test fails correctly.

4. Running playwright or via folio (against HTTPS):
```
~> npx folio --param browserName=chromium

Running 1 test using 1 worker
  1 passed (2s)
```
This test should have failed, as the proxy is bogus. But in this case, I'm using an HTTPS address (e.g. `https://www.saucedemo.com/`).

**Comment by mxschmitt** [comment]
Please try to set the environment variable names in lower-case. Firefox does only support lower-case http proxy from the environment.

```
export http_proxy=http://localhost:1234
export https_proxy=https://localhost:1234
```

**Comment by alexplischke** [comment]
My above examples were run with `chromium` though 💡

**Comment by mxschmitt** [comment]
Good catch :+1:, the last one I saw and also the one before and so I was confused. So currently the browsers also need access to DNS to work with proxies, could you try accessing an IP directly like `https://1.1.1.1`?
int

(Background: I recently setup an infrastructure where I run Playwright inside containers which are completely isolated and only have access to DNS configured. They interact with the internet over a http forwarding proxy which is configured using the lower-case environment variables. There I also ran into the issue that firefox was not working, digged into it and found out that firefox only supports the lower-case variants.)

**Comment by alexplischke** [comment]
> could you try accessing an IP directly like https://1.1.1.1?

Yup, it is!

The run environment is actually a local mac (laptop). No docker involved.
The interesting thing is that I can't even replicate the exact same example I showed yesterday.

Having said that, setting HTTP_PROXY/HTTPS_PROXY (upper/lowercase) as env variables does not seem to take any effect whatsover. Whether I set those envs and launch the browser from the commandline or through playwright (unless I use CLI args `/Applications/Google\ Chrome.app/Contents/MacOS/Google\ Chrome --proxy-server "http://localhost:1234"`, which _does_ work).

When using playwright, specifying launch options is the only thing that makes it work: `await playwright['chromium'].launch({proxy: {server: process.env.HTTP_PROXY}});`
Would be great if I could pass those options directly via the folio CLI args as well 🙏

**Comment by mxschmitt** [comment]
Oh I know the problem now! Each browser implements the logic for "fetching the proxy from the operating system" differently. On Windows its in the global OS settings, for macOS its in the global network settings and for Linux its fetched via the environment variables.

That's why its working for me without any issues, because I used it in Docker and by that with Linux/Ubuntu.

Not sure if we should create an issue for it to standardize the proxy fetching logic via the environment variables to make this work on all the operating systems. Feel free to raise an issue upstream on microsoft/playwright to request this feature there.

Regarding the folio specific issue, it would be possible to overwrite the launch options and add the proxy there as a workaround. See here: https://github.com/microsoft/playwright-test#modify-launch-options

**Comment by alexplischke** [comment]
🎉 😄 Thanks for the info!

> Regarding the folio specific issue, it would be possible to overwrite the launch options and add the proxy there as a workaround. See here: https://github.com/microsoft/playwright-test#modify-launch-options

We have a bit of a special usecase, so this may not be available to us 😓 

In our case, we effectively provide a wrapper around playwright-folio (https://github.com/saucelabs/saucectl & https://github.com/saucelabs/sauce-playwright-runner). So it's the user that provides the test files, but it's on us to set the proxy automatically depending on the environment it is executed on.

Effectively this means there is two ways for us to do it:
a) (Potentially) Via ENV vars (once/if they work the same way across different operating systems 😄 )
b) Via the folio CLI arg (since the CLI is what we invoke, so we have direct control over the params here).

**Comment by mxschmitt** [comment]
I think once the folio rewrite is done (currently in the making) then it should be easy to do that via folio. I will talk about that topic (http_proxy env var as default on each OS) tomorrow in our sync and let you know.

**Comment by alexplischke** [comment]
@mxschmitt Just wanted to check if there's been any updates? :)

**Comment by mxschmitt** [comment]
I didn't had a chance yet, since Friday was a day-off. For now I would recommend to wait for the new folio release, which should be ready in a few weeks and/or create an upstream feature request that this feature can be considered. Thanks!

**Comment by mxschmitt** [comment]
Closing as part of the triage process since the [new test-runner](https://playwright.dev/docs/test-intro) got released yesterday which is quite different. Please create a new issue if it still does not work for you.
